### PR TITLE
LPS-158259 Set scopes to avoid 401 Unauthorized error when using OAuth 2.0 as the authorization mechanism in GraphQL

### PR DESCRIPTION
**Pull request by ccorreagg**
This PR contains the code to allow using OAuth 2.0 as the authorization mechanism to interact with GraphQL: https://issues.liferay.com/browse/LPS-158259.

For each companyId, a java Servlet is created to manage the requests through GraphQL. The servlet is generated by using [this external library](https://www.graphql-java.com/), configuring its schema properly.

Each GraphQL request is transformed into this two options:

-  The similar request in the JAX-RS REST API layer, as every JAX-RS application created with REST Builder is exposed automatically in GraphQL. That is made by REST Builder by implementing the `ServletData` interface (throught the creation of the class `ServletDataImpl` in each REST API).
- A `GraphQLDTOContributor` invocation, based on the type of operation desired (CREATE, DELETE, etc)

As the external library uses the `DataFetcher` interface to execute the desired code for each request by invocating the method `get(DataFetchingEnvironment var1)`, the scope validations might be included before executing the JAX-RS target method, as in that point, the request body has been properly parsed by the library and the context has been set.

**Current validations**

The JAX-RS scope validations are being already performed in the following classes:

- `HttpMethodFeature`: Validates the http method of the request.
- `AnnotationFeature`: Validates the scopes configured through certain annotations in the JAX-RS resource method.

To use the same validations in the GraphQL layer, those functionality has been refactored by moving it to an implementations of the new interface `JaxRsResourceScopeChecker` with the method `public boolean check(Class<?> resourceClass, Method resourceMethod)`.

**GraphQL scope validations**

To validate the scope, a new interface has been created in Vulcan: `GraphQLRequestContextValidator`. The point of that new inteface is to provide a way to validate the `GraphQLRequestContext`, that must include the context of the GraphQL Request based on the necessary parameters. That validation is performed in the `BaseDataFetcher` class.

The context `GraphQLRequestContext` will be build depending of the type of request:

- The request that is delegated to the ServletData implementation is generated with the class `ServletDataRequestContext`.
- The request that is delegated to the GraphQLDTOContributor implementation is generated with the class `GraphQLDTOContributorRequestContext`.

A new implementation of the `GraphQLRequestContext` interface has been included to provide a way to check the GraphQL context by using the same mechanisms that in REST; that means by using the refactored `JaxRsResourceScopeChecker` implementations.

**NOTE:** The new methods created in the `GraphQLDTOContributor` and `ServletData` classes have been added to get the equivalent JAX-RS resource to the operation that is being performed. It is not the beautiest solution, but if any kind of interface was included into Vulcan API, that would be a breaking change for the customers that are using REST Builder to create its own APIs without updating Vulcan API package.

### CODE REVIEW

**Comment by ccorreagg** [comment]
ci:test:relevant

**Comment by ccorreagg** [comment]
ci:test:relevant

**Comment by magjed4289** [comment]
Failures in ci:test:security are not related to the PR. ✅ 
Falires in ci:test:relevant are not related to the PR. ✅ 
OK to manually forward.

**Comment by arthurchan35** [comment]
ci:test:relevant

File: modules/apps/oauth2-provider/oauth2-provider-scope-impl/src/main/java/com/liferay/oauth2/provider/scope/internal/graphql/OAuth2GraphQLRequestContextValidator.java [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+			FrameworkUtil.getBundle(graphQLRequestContext.getResourceClass()));
+		_scopeContext.setCompanyId(graphQLRequestContext.getCompanyId());
+
+		if (OAuth2ProviderScopeLiferayAccessControlContext.
+				isOAuth2AuthVerified()) {
+
+			_enableSAP(graphQLRequestContext.getApplicationName());
+
+			if (!_scopeChecker.checkScope(
+					graphQLRequestContext.getHttpMethod())) {
```
Comment (stian-sigvartsen) [inline modules/apps/oauth2-provider/oauth2-provider-scope-impl/src/main/java/com/liferay/oauth2/provider/scope/internal/graphql/OAuth2GraphQLRequestContextValidator.java:74]: This assumes that the JAX-RS application is using HTTP verbs for scopes (`HttpMethodFeature`). This is the default for JAX-RS applications, and also what REST Builder generated ones use. However we should support other ones also. Most notably annotation based scopes `AnnotationFeature` .

**Comment by ccorreagg** [comment]
After making tests with more than one GraphQL request in the same request body, I modified the current code to set always the depth to 1 as there could be more than one validation for the same request and several scopes could be involved. The code has been merged in the same commit: https://github.com/liferay-appsec/liferay-portal/pull/846/commits/8d5116b18e52649d8bf64fa887459e281f6f1696

**Comment by ccorreagg** [comment]
ci:test:sf

**Comment by ccorreagg** [comment]
ci:test:stable

**Comment by ccorreagg** [comment]
ci:test:sf

**Comment by ccorreagg** [comment]
ci:test:sf

**Comment by ccorreagg** [comment]
ci:test:security

**Comment by ccorreagg** [comment]
Rebased to remove conflicts and also to include the GraphQL library version update changes into the current branch

**Comment by ccorreagg** [comment]
ci:test:security

**Comment by ccorreagg** [comment]
ci:test:sf

**Comment by ccorreagg** [comment]
ci:test:relevant

**Comment by ccorreagg** [comment]
ci:test:sf

**Comment by ccorreagg** [comment]
ci:test:relevant

**Comment by ccorreagg** [comment]
ci:test:security

**Review by stian-sigvartsen — CHANGES_REQUESTED** [review]
This is very close now I think. I made some slight changes in liferay-appsec/liferay-portal/pull/869

- The proposed SPI name JaxRsResourceScopeChecker sounds too similar to the existing ScopeChecker SPI. Actually it sounds like an implementation of that SPI
- Some regressions in OAuth 2 Provider's behaviour and configurability (I've made a mental note to add more tests!)

**Comment by stian-sigvartsen** [comment]
@ccorreagg One additional thing ... The PR means `oauth2-provider-impl` depends on the `portal-vulcan-api` module's package `com.liferay.portal.vulcan.graphql.validation` . We may consider moving that dependency to a separate module, either on the Vulcan or the OAuth 2 provider side.

**Comment by stian-sigvartsen** [comment]
We merged this across two tickets. The 2nd part is ccorreagg/liferay-portal/pull/13

**Pull request closed without merging**